scapy/contrib/igmp.py

```diff
@@ -55,134 +56,80 @@ class IGMP(Packet):
-               a proper IGMP packet.
-      False   One of more validation checks failed and no fields 
-               were adjusted.
-
-    The function will examine the IGMP message to assure proper format. 
-    Corrections will be attempted if possible. The IP header is then properly 
-    adjusted to ensure correct formatting and assignment. The Ethernet header
-    is then adjusted to the proper IGMP packet format.
-    """
-
-# The rules are:
-#   1.  the Max Response time is meaningful only in Membership Queries and should be zero 
-#       otherwise (RFC 2236, section 2.2)
-
-    if (self.type != 0x11):         #rule 1
-      self.mrtime = 0
-      
-    if (self.adjust_ip(ip) == True):
-      if (self.adjust_ether(ip, ether) == True): return True
-    return False
-
-#--------------------------------------------------------------------------
-  def adjust_ether (self, ip=None, ether=None):
-    """Called to explicitly fixup an associated Ethernet header
-
-    The function adjusts the ethernet header destination MAC address based on 
-    the destination IP address.
-    """
-# The rules are:
-#   1. send to the group mac address address corresponding to the IP.dst
-    if ip != None and ip.haslayer(IP) and ether != None and ether.haslayer(Ether):
-      iplong = atol(ip.dst)
-      ether.dst = "01:00:5e:%02x:%02x:%02x" % ( (iplong>>16)&0x7F, (iplong>>8)&0xFF, (iplong)&0xFF )
-      # print "igmpize ip " + ip.dst + " as mac " + ether.dst 
-      return True
-    else:
-      return False
-
-#--------------------------------------------------------------------------
-  def adjust_ip (self, ip=None):
-    """Called to explicitly fixup an associated IP header
-
-    The function adjusts the IP header based on conformance rules 
-    and the group address encoded in the IGMP message.
-    The rules are:
-    1. Send General Group Query to 224.0.0.1 (all systems)
-    2. Send Leave Group to 224.0.0.2 (all routers)
-    3a.Otherwise send the packet to the group address
-    3b.Send reports/joins to the group address
-    4. ttl = 1 (RFC 2236, section 2)
-    5. send the packet with the router alert IP option (RFC 2236, section 2)
-    """
-    if ip != None and ip.haslayer(IP):
-      if (self.type == 0x11):
-        if (self.gaddr == "0.0.0.0"):
-          ip.dst = "224.0.0.1"                   # IP rule 1
-          retCode = True                     
-        elif isValidMCAddr(self.gaddr):
-          ip.dst = self.gaddr                    # IP rule 3a
-          retCode = True
+                    _IGMP_Gaddr_IPField("gaddr", "0.0.0.0")]
+
+    def post_build(self, p, pay):
+        """Called implicitly before a packet is sent to compute and place IGMP checksum.
+
+        Parameters:
+          self    The instantiation of an IGMP class
+          p       The IGMP message in hex in network byte order
+          pay     Additional payload for the IGMP message
+        """
+        p += pay
+        if self.type != 0x11:
```

Comment:
First, this code should probably be commented. But more importantly, this is not how it should work. With Scapy, you should be able to send a packet that does not match the specification. What you may want to do is that if the user leaves the default value for the `mrtime` field, **then** it should be set to `20` when the type is `11`, and to `0` otherwise.

Or maybe we leave the default to 0 and the user has to change its value when needed.

Reply:
You're right. I should have left igmpize